### ISSUE TITLE
execution order of plugins initialization & player initialize callback function

### ISSUE THREAD

**Issue by Vuuo** [videojs/video.js#2505]
videojs version: 4.11

I wanna make a plugin. The plugin will attach a function to the player which get initialized. Then I wanna be able to call the attached function in the player initialize callback function.

```javascript
// my_plugin.js
(function(window, videojs) {
  my_plugin = function(init_options) {
    return console.log('my_plugin initialized');
    this.my_func = function() {
      return console.log('my_func invoked');
    }
  };
  return videojs.plugin('my_plugin', my_plugin);
})(window, window.videojs);
```

```javascript
// index.html
...
<script>
  player = videojs('player', {
    techOrder: ["flash", "html5"]
    ...
    plugins: {
      my_plugin: {}
    }
  }, function(){
    console.log('player initialize callback function invoked');
    this.my_func();
  });
</script>
```

If the tech is `flash`, `this.my_func()` will get executed well. However if the tech is `html5`, it throws error saying that the function is not defined.

Further investigation tells me that in `flash` mode, the plugin get initialized before the callback function runs. While in `html5` mode, the callback function get executed before the plugin initialization.

Looks like the behaviour is inconsistent across flash and html5. Or am I doing something wrong?
Any workaround would also be great.

Thanks

**Comment by mmcc** [comment]
This looks like it is a bug in 4.x, but there is very little chance we're going to put out a patch to fix this in 4.12. The easiest workaround would me to call your function in your ready block instead of using options to initialize:

```javascript
var player = videojs('player', {}, function() {
  this.my_plugin();
  this.my_func();
});
```

I was hoping this had been fixed in 5.0 already, but it looks like we've got a [new issue](2510) to fix regarding plugin initialization via an options object. So basically, same workaround required in master right now, but we're working on it.

All of this being said, if you're planning on releasing this plugin into the wild, please add things to the videojs prototype with extreme caution. There's the simple case of two plugins potentially overwriting each other, but we also try to subscribe to the "don't modify objects you don't own" philosophy, so generally we suggest adding new methods under your plugin rather than directly to vjs. Feel free to totally ignore all of that if you plan on keeping this plugin to yourself; this is more applicable if you plan on releasing the plugin publicly.

**Comment by mmcc** [comment]
I initially accidentally closed this, but I think #2510 probably covers this in terms of tracking. Closing this in lieu of that issue, but feel free to continue the conversation here or there.

**Comment by Vuuo** [comment]
I see. Thank you very much.
